### ISSUE TITLE
Potential bug in Mixture

### ISSUE THREAD

**Issue by viveksck** [bayespy/bayespy#48]
I am trying to implement a simple mixture of Categorical Distributions where
When p = 0 I need [0,0] 
and when p = 1 I need [1,1] 

This can easily be achieved through a mixture model as follows: Component Distribution when p =0 should be Categorical([[1,0],[1,0]]) and Component Distribution when p = 1 should be Categorical([[0,1],[0,1]])

p is the selector which is Categorical([0.5,0.5])

However this simple example does not work. It generates a sample [0,1] which should never occurr. 

I have verified that individual components generate the right samples. Just trying to use Mixture to select each component however does not work.
.
```python
from bayespy.nodes import Categorical, Mixture
from collections import Counter
p = Categorical([0.5,0.5])

X = np.array([[[0.999,0.001],[0.999,0.001]],[[0.001,0.999],[0.001,0.999]]])

z.random()
Output: array([0, 1]) << We should never actually see (0,1) because when we select a component it should only either yield [0,0] or [1,1]

q = Categorical(X[0], plates=(1000,2))  << This works only [0,0] should be returned
array([[0, 0],
       [0, 0],
       [0, 0],
       ..., 
       [0, 0],
       [0, 0],
       [0, 0]])

Similarly X[1] Categorical(X[1], plates=(1000,2)) only returns [1,1]

**Comment by jluttine** [comment]
Thanks for the report!

However, the example code doesn't run because `z` is not defined. Also, I can't see any use of `Mixture` node. Could you provide a simple example which gives a different output than you expected and tell what you would have expected to see?

I don't see why would you need to just copy the outcome of the parent deterministically to be the value of the child node. Probably there is a better way to achieve what you want, so you could elaborate a little bit on what it is that you are trying to do.

But as far as I understand, this probably answers your question: Note that if you use `random()` method for a node, it is not a true generative sampling process. It draws a sample from the current state of the VB factor for that node *independently* of its parents. So, as `p` has probabilities `[0.5, 0.5]`, the child node has the same probabilities for each element because of the deterministic mapping. Thus, you'll end up sampling the elements of the child node from `[0.5, 0.5]`. Note, the key point is that the sampling is done independently of the parent because the distributions of the nodes are independent of other nodes. Thus, it is not a true generative sample.

And a rule of thumb: Don't use stochastic nodes like `Mixture` to create deterministic (or nearly deterministic) mappings: You'll end up having extremely strongly coupled variables in the posterior but they are approximated to be independent in the VB posterior, so the approximation is most likely extremely bad.

I hope this helps. Please don't hesitate to ask further questions!

**Comment by jluttine** [comment]
Was this problem solved (i.e., did my answer explain the weird things you saw) or could you provide a simple example to reproduce the bug at some point so I could take a look?

**Comment by viveksck** [comment]
Thanks! Yes I initially tried to use mixture to ensure that a set of categorical variables have the same value dependent on a parent.  So if p = 0, I wanted a vector variable all to take [0,0]. If p = 1, I wanted that to take [1,1]. Your comment above indicates that this is not a right way to model and so yes you may close this issue.

Just for clarity, here is my code (apologies for the copy paste error) that I had initially tried:

```python
from bayespy.nodes import Categorical, Mixture
from collections import Counter
import numpy as np
p = Categorical([0.5,0.5])
X = np.array([[[0.999,0.001],[0.999,0.001],[0.999,0.001]],[[0.001,0.999],[0.001,0.999], [0.001,0.999]]])
z = Mixture(p, Categorical, X, cluster_plate=-2)
z.random()Output: array([1, 1, 0]) << We should never actually see (1,1,0) because when we select a component it should only either yield [0,0,0] or [1,1,1]

Basically if p =0 , then the first component of the mixture is chosen and so when I draw a sample from it, I expect z to be [0,0,0]. Similarly if p happened to be 1, then I expect z to be [1,1,1]

q = Categorical(X[0], plates=(100,3)  << This works only [0,0,0] should be returned
Similarly X[1] Categorical(X[1], plates=(100,3)) only returns [1,1,1]

**Comment by jluttine** [comment]
Ok, thanks!
